### ISSUE TITLE
BrokenPipeError with big containers (HTTPS only)

### ISSUE THREAD

**Issue by victorsndvg** [singularityhub/sregistry#108]
Hi @vsoch ,

try to push a 2.6Gb container with my https token and I get the error below. If  I change the protocol to http in my sregistry token (`$HOME/.sregistry`) it works.

Any idea to fix this?

Thanks!

```
sregistry push --name mso4sc/bigfile --tag test salome-8.3.0.simg 
[client|registry] [database|sqlite:////root/.singularity/sregistry.db]
Traceback (most recent call last): 0/2658 MB - 00:00:00
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 601, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 357, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1065, in _send_output
    self.send(chunk)
  File "/usr/local/lib/python3.6/http/client.py", line 986, in send
    self.sock.sendall(data)
  File "/usr/local/lib/python3.6/ssl.py", line 972, in sendall
    v = self.send(byte_view[count:])
  File "/usr/local/lib/python3.6/ssl.py", line 941, in send
    return self._sslobj.write(data)
  File "/usr/local/lib/python3.6/ssl.py", line 642, in write
    return self._sslobj.write(data)
BrokenPipeError: [Errno 32] Broken pipe

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 440, in send
    timeout=timeout
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 639, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/local/lib/python3.6/site-packages/urllib3/util/retry.py", line 357, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.6/site-packages/urllib3/packages/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 601, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 357, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1065, in _send_output
    self.send(chunk)
  File "/usr/local/lib/python3.6/http/client.py", line 986, in send
    self.sock.sendall(data)
  File "/usr/local/lib/python3.6/ssl.py", line 972, in sendall
    v = self.send(byte_view[count:])
  File "/usr/local/lib/python3.6/ssl.py", line 941, in send
    return self._sslobj.write(data)
  File "/usr/local/lib/python3.6/ssl.py", line 642, in write
    return self._sslobj.write(data)
urllib3.exceptions.ProtocolError: ('Connection aborted.', BrokenPipeError(32, 'Broken pipe'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/sregistry", line 11, in <module>
    load_entry_point('sregistry', 'console_scripts', 'sregistry')()
  File "/usr/local/lib/sregistry-cli/sregistry/client/__init__.py", line 352, in main
    subparser=subparsers[args.command])
  File "/usr/local/lib/sregistry-cli/sregistry/client/push.py", line 42, in main
    tag=args.tag)
  File "/usr/local/lib/sregistry-cli/sregistry/main/registry/push.py", line 88, in push
    r = requests.post(url, data=monitor, headers=headers)
  File "/usr/local/lib/python3.6/site-packages/requests/api.py", line 112, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 490, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', BrokenPipeError(32, 'Broken pipe'))
```

**Comment by vsoch** [comment]
What do you mean "if you change the protocol?"

**Comment by victorsndvg** [comment]
jaja

I mean writing the token like this:
```
{ "registry": { "token": "...", "username": "...", "base": "http://..." } }
```
instead of:
```
{ "registry": { "token": "...", "username": "...", "base": "https://..." } }
```

**Comment by vsoch** [comment]
I think this is a large, known bug that I can't do much about other than suggest that you use http, and wiat for this to be addressed --> https://github.com/requests/requests/issues/2422

**Comment by Trophime** [comment]
Out of curiosity what is the actual limit supported by https?

**Comment by vsoch** [comment]
I think (generally speaking) 2GB? https://stackoverflow.com/questions/2659952/maximum-length-of-http-get-request Oh here is an idea! What if you set it to 0 - that is supposed to disable it?

![image](https://user-images.githubusercontent.com/814322/37405138-45b9e6c6-276a-11e8-8564-07c045c9956e.png)

from here: http://nginx.org/en/docs/http/ngx_http_core_module.html

I think the more elegant fix would be to change the way the upload happens, period, I remember @dctrud mentioned something about this, but I can't remember the better way to do it :P

**Comment by dctrud** [comment]
I have used this... https://github.com/juliomalegria/django-chunked-upload

**Comment by vsoch** [comment]
Awesome thank you @dctrud ! I'm going to open that up as its own issue because it's something that will improve the registry across the board I think.

**Comment by victorsndvg** [comment]
Great! @vsoch and @dctrud  

It would be nice to have this fixed! :+1:

**Comment by vsoch** [comment]
PR has been merged, and chunked uploads added #134
